### PR TITLE
kube-apiserver: add --assume-pruned-custom-resource-metadata-in-storage to optimize read-path

### PR DESCRIPTION
**Pull request by sttts**
Due to missing ObjectMeta pruning we have a marshal/unmarshal orgy going on in the CR read-path:

  1. unmarshal JSON string from etcd into `obj := map[string]interface{}{}`
  2. marshal `obj[metadata]` back into a JSON string
  3. a) unmarshal JSON string into `metadata := meta.ObjectMeta{}` – this prunes
     b) unmarshal field by field in case the complete `json.Unmarshal` fails – this drops invalid fields, but is even slower. Usually only a) runs.
  4. marshal `metadata` to JSON string
  5. unmarshal JSON string into `metadata := map[string]interface{}{}`
  6. `obj[metadata] = metadata`
  7. marshal `obj` into JSON string an send to the client

Since Kubernetes 1.11 we have metadata pruning in place for new(ly written) CustomResources. Hence, we could assume that pruning is unnecessary, and hence speed up our CR read path considerably.

Note: there is one catch if you have embedded objects in your CRD schema. These technically can always be unpruned in etcd because you don't know which schema was used when writing them. When setting `x-kubernetes-embedded-object: true` at some point, data in etcd might be invalid.

/kind cleanup

```release-note
kube-apiserver: add --assume-pruned-custom-resource-metadata-in-storage to optimize read-path
```

### CODE REVIEW

**Comment by sttts** [comment]
/assign @liggitt @jpbetz 

@jpbetz do you have some CR benchmark from your binary encoding tests that you could run against this PR ?

**Comment by sttts** [comment]
@wojtek-t do we have CRD benchmarks we could run against this?

File: cmd/kube-apiserver/app/options/options.go [hunk 1]
```diff
@@ -93,6 +93,12 @@ type ServerRunOptions struct {
 	ServiceAccountTokenMaxExpiration time.Duration
 
 	ShowHiddenMetricsForVersion string
+
+	// AssumePrunedCustomResourceMetaDataInStorage set to true optimizes the read code-path for CRs
+	// assuming that ObjectMeta has been pruned when writing the objects. This can be
+	// assumed for every cluster installed with or after 1.11 when ObjectMeta pruning
+	// was added.
+	AssumePrunedCustomResourceMetaDataInStorage bool
```
Comment (liggitt) [inline cmd/kube-apiserver/app/options/options.go:101]: would it be terrible to default this to deriving the setting per-CRD from the creation timestamp of the CRD? 

for example, if a CRD's creation timestamp was well after the 1.10 EOL date (https://github.com/kubernetes/website/blob/main/content/en/releases/patch-releases.md#non-active-branch-history), like 2020-01-01

**Comment by wojtek-t** [comment]
> @wojtek-t do we have CRD benchmarks we could run against this?

@sttts - unfortunately not - we're missing reasonable CRD-related scale/performance related tests.
I agree it's important to get them, but this have never happened yet.

**Comment by dims** [comment]
Is this PR still needed, please rebase if so (or we can close it?)

**Comment by dims** [comment]
This PR is older than 4 weeks and needs as rebase.

Please rebase the PR against latest master and reopen if still needed.

/close

**Pull request closed without merging**